records/2024-08-14.suspend-resume-callbacks.md

```diff
@@ -0,0 +1,120 @@
+# Suspension and resumption of service callbacks
+
+Date: 2024-08-14
+
+## Status
+
+**DRAFT**.
+
+
+## Context
+
+_TODO: Describe context that triggered the writing of this ADR._
+
+Notify offers the delivery of notification callbacks, containing info such as notification delivery status, recipient, and delivery date, for services who choose to configure them. When a callback fails to be delivered to the service's callback API, Notify will attempt to resend the callback up to 5 times, with a 5 seconds retry back off, before dropping a callback all together.
+
+The drawback of this retry mechanism is that Notify can experience spikes in processing demands in certain scenarios. Since 100% uptime of a service's callback API cannot be guaranteed, when a service sends a large number of notifications while their API is down this it results in a spike in processing demands.
+
+For example, if a service sends 1000 notifications while their API is down then Notify is responsible for fulfilling 5000 requests to that service's callback API. As Notify grows, both in the number of active services and total notification throughput, so will the processing demand associated with fulfilling callbacks.
+
+This ADR introduces mitigation strategies to reduce system load in the scenario where a callback service is down, and an associated Notify service is continuing to send notifications.
+
+## Options
+
+## Trigger alarms to prompt manual suspension of a service's callbacks
+
+### Alarms
+[Some alarms have already been implemented.](https://github.com/cds-snc/notification-terraform/commit/40f105da8df7e4cf89ae418bef4be96e888f00b4)
+
+Thresholds subject to change.*
+
+#### Warning Alarm
+
+Callbacks fail consistent delivery, requiring retries before being delivered successfully.
+- **5 notification callbacks** failed delivery within **15 minutes**
+
+#### Critical Alarm
+
+Callbacks reach max retries and fail delivery entirely.
+- **5 notification callbacks** fail, exceeding the max retries of five within **30 minutes**.
```

Comment:
I would make this 10 notification callbacks

Reply:
For larger sends, I agree this could be too sensitive. In related work we're reducing the callback request timeout from `5 seconds` to `1 second`, which will make this more sensitive as we will hit the max retries quicker.